### PR TITLE
Support workspaces.

### PR DESCRIPTION
**Pull request by kjvalencik**
Implements https://github.com/neon-bindings/neon/issues/297

### CODE REVIEW

File: cli/src/project.ts [hunk 1]
```diff
@@ -3,24 +3,41 @@ import Crate from './crate';
...
+  private constructor(root: string, options: ProjectOptions) {
+    let { crate, targetDirectory } = options;
     this.root = root;
+    this.targetDirectory = targetDirectory!;
     this.crate = new Crate(this, { subdirectory: crate });
   }
 
+  static async create(root: string, options: Partial<ProjectOptions> = {}): Promise<Project> {
+    let { crate = 'native' } = options;
+    const { stdout } = await execFile("cargo", ["metadata", "--format-version=1"], {
```
Comment (kjvalencik) [inline cli/src/project.ts]: This is only necessary in `build` and would be fewer changes if it was moved there, but it felt wrong not to have `targetDirectory` on the `Project`.

Comment (dherman) [reply cli/src/project.ts]: Seems 👍  to me.

File: crates/neon-runtime/build.rs [hunk 2]
```diff
@@ -26,6 +26,11 @@ fn main() {
 
     // 4. Copy native build artifacts
     copy_build_artifacts(&native_to, &out_dir);
+
+    // Rebuild if any of these environment variables change
```
Comment (dherman) [inline crates/neon-runtime/build.rs]: Can you help me understand why this was needed for this PR?

**Comment by dherman** [comment]
Actually one more big picture question: what exactly do we mean by a workspace here? Like, `<root>/native` can contain a workspace with multiple crates? Or do we mean somehow nesting a whole Neon project (including the JS) inside an outer cargo workspace? Or does this support both?

Also, it would be good to think about whether we want `neon new` to support any other project structure than the current one. Resolving that question doesn't block this PR, but worth discussing.

**Comment by kjvalencik** [comment]
@dherman Technically, this means supporting projects where the compiled binary isn't in `native/target`. Practically, that means when it's part of a workspace. That could mean any of the following:

* Part of a workspace with multiple neon modules
* Part of a workspace with multiple crates, one being a neon module
* A top level `Cargo.toml` that makes `native/` the only member of the workspace (fixes VS Code and allows easier `cargo test`)

I think it would be nice for the default `neon new` to drop a `Cargo.toml` with `native` in the workspace, but I think that can be a different PR.

File: cli/src/project.ts [hunk 3]
```diff
@@ -3,24 +3,41 @@ import Crate from './crate';
...
   readonly root: string;
+  readonly targetDirectory: string;
   readonly crate: Crate;
 
-  constructor(root: string, options: ProjectOptions = {}) {
-    let { crate = 'native' } = options;
+  private constructor(root: string, options: ProjectOptions) {
+    let { crate, targetDirectory } = options;
     this.root = root;
+    this.targetDirectory = targetDirectory!;
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply cli/src/project.ts]: It's not; it's from an older version of the code. I need to remove it.

File: crates/neon-runtime/build.rs (hunk 2 above)
Comment (kjvalencik) [reply crates/neon-runtime/build.rs]: It's an existing bug that this PR made apparent. These environment variables impact the compatibility of `libneon`. For example, with these changes `test/dynamic` compiles `neon-runtime` and then `test/electron` tries to use it and fails.

**Comment by dherman** [comment]
I think changing the default layout should have an RFC to explain and motivate the design change. I definitely _love_ the idea of having `cargo` work in the project root. I drafted [one RFC that proposed changing the directory layout](https://github.com/neon-bindings/rfcs/pull/4) although I think some of the particular ideas there missed the mark. I wonder how far cargo and/or Neon would need to evolve before we could let people use vanilla cargo without needing a wrapper CLI tool.

**Pull request merged**